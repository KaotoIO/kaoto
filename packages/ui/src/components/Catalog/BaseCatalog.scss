.catalog-list {
  overflow: auto;
  padding: var(--pf-v5-global--spacer--sm);
}
