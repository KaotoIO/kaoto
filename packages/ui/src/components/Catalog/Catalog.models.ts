export interface ITile {
  type: string;
  name: string;
  title: string;
  description?: string;
  headerTags?: string[];
  tags: string[];
  version?: string;
  provider?: string;
}

export type TileFilter = (item: ITile) => boolean;

export const enum CatalogLayout {
  Gallery = 'Gallery',
  List = 'List',
}
