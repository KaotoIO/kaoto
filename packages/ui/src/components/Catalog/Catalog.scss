.catalog {
  overflow: inherit;
  max-height: 100%;
  display: flex;
  flex-direction: column;

  &__filter {
    margin-bottom: var(--pf-t--global--spacer--md);
  }

  &__base {
    margin-bottom: var(--pf-t--global--spacer--md);
  }
}
