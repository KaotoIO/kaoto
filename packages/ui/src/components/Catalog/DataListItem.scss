.catalog-data-list-item {

    &__title-div-left {
        &__title {
            font-family: var(--pf-v5-global--FontFamily--heading);
            margin-right: 5px;
            margin-left: 5px;
        }

        &__icon {
            width: 25px;
            height: 25px;
            vertical-align: middle;
        }
    }

    &__title-div-right {
        display: flex;
        flex-direction: row;
        justify-content: left;
        flex-flow: wrap;
    }

    @media (min-width: 768px) {
        &__title-div-right {
            justify-content: right;
        }
    }

    &__description {
        overflow: hidden;
        display: -webkit-box;
        -webkit-line-clamp: 2;
        -webkit-box-orient: vertical;
    }
}
