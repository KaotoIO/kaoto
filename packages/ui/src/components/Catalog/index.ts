export * from './Catalog';
export * from './CatalogLayoutIcon';
export * from './Catalog.models';
export * from './DataListItem';
export * from './Tile';
export * from './tag-color-resolver';
