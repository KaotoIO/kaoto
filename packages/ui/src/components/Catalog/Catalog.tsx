import { FunctionComponent, PropsWithChildren, useCallback, useMemo, useState } from 'react';
import { useDebounceValue } from 'usehooks-ts';
import { useLocalStorage } from '../../hooks';
import { LocalStorageKeys } from '../../models';
import { BaseCatalog } from './BaseCatalog';
import { CatalogLayout, ITile } from './Catalog.models';
import './Catalog.scss';
import { CatalogFilter } from './CatalogFilter';
import { filterTiles } from './filter-tiles';
import { sortTags } from './sort-tags';

interface CatalogProps {
  /** Tiles list */
  tiles: ITile[];

  /** Callback being called when a Tile is clicked */
  onTileClick?: (tile: ITile) => void;
}

export const Catalog: FunctionComponent<PropsWithChildren<CatalogProps>> = (props) => {
  const [activeLayout, setActiveLayout] = useLocalStorage(LocalStorageKeys.CatalogLayout, CatalogLayout.Gallery);

  /** Selected Group */
  const [searchTerm, setSearchTerm] = useDebounceValue('', 500, { trailing: true });
  const [filterTags, setFilterTags] = useState<string[]>([]);

  /** All tags, sorted to have selected and prioritized tags first */
  const { sortedTags, overflowIndex } = useMemo(() => {
    return sortTags(props.tiles, filterTags);
  }, [filterTags, props.tiles]);

  /** All providers */
  const providers = useMemo(() => {
    const providersSet = new Set<string>();
    props.tiles.forEach((tile) => {
      providersSet.add(tile.provider ?? 'Community');
    });

    return Array.from(providersSet);
  }, [props.tiles]);

  /** Selected Providers */
  const [selectedProviders, setSelectedProviders] = useState<string[]>(providers);

  const filteredTiles = useMemo(() => {
    return filterTiles(props.tiles, { searchTerm, searchTags: filterTags, selectedProviders });
  }, [filterTags, props.tiles, searchTerm, selectedProviders]);

  /** Set the tiles groups */
  const tilesGroups = useMemo(() => {
    const groups: Record<string, ITile[]> = {};
    filteredTiles.forEach((tile) => {
      if (!groups[tile.type]) {
        groups[tile.type] = [];
      }
      groups[tile.type].push(tile);
    });
    return Object.entries(groups).map(([group, tiles]) => ({ name: group, count: tiles.length }));
  }, [filteredTiles]);

  const [activeGroups, setActiveGroups] = useState<string[]>(tilesGroups.map((g) => g.name));

  const filteredTilesByGroup = useMemo<ITile[]>(() => {
    return filteredTiles.filter((tile) => activeGroups.includes(tile.type));
  }, [activeGroups, filteredTiles]);

  const onFilterChange = useCallback(
    (_event: unknown, value = '') => {
      setSearchTerm(value);
    },
    [setSearchTerm],
  );

  const onTileClick = useCallback(
    (tile: ITile) => {
      props.onTileClick?.(tile);
    },
    [props],
  );

  const onTagClick = useCallback((_event: unknown, value = '') => {
    setFilterTags((previousFilteredTags) => {
      return previousFilteredTags.includes(value) ? previousFilteredTags : [...previousFilteredTags, value];
    });
  }, []);

  const onSelectProvider = useCallback(
    (provider: string) => {
      setSelectedProviders(
        selectedProviders.includes(provider)
          ? selectedProviders.filter((selection) => selection !== provider)
          : [provider, ...selectedProviders],
      );
    },
    [selectedProviders],
  );

  return (
    <div className="catalog">
      <CatalogFilter
        className="catalog__filter"
        searchTerm={searchTerm}
        groups={tilesGroups}
        providers={providers}
        selectedProviders={selectedProviders}
        tags={sortedTags}
        tagsOverflowIndex={overflowIndex}
        layouts={[CatalogLayout.Gallery, CatalogLayout.List]}
        activeGroups={activeGroups}
        activeLayout={activeLayout}
        filterTags={filterTags}
        onChange={onFilterChange}
        setActiveGroups={setActiveGroups}
        setActiveLayout={setActiveLayout}
        setFilterTags={setFilterTags}
        onSelectProvider={onSelectProvider}
      />
      <BaseCatalog
        className="catalog__base"
        tiles={filteredTilesByGroup}
        catalogLayout={activeLayout}
        onTileClick={onTileClick}
        onTagClick={onTagClick}
      />
    </div>
  );
};
