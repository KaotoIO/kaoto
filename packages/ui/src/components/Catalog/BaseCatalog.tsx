import { DataList, Gallery, Pagination } from '@patternfly/react-core';
import { FunctionComponent, useCallback, useEffect, useRef, useState } from 'react';
import './BaseCatalog.scss';
import { CatalogLayout, ITile } from './Catalog.models';
import { CatalogDataListItem } from './DataListItem';
import { Tile } from './Tile';

interface BaseCatalogProps {
  className?: string;
  tiles: ITile[];
  catalogLayout: CatalogLayout;
  onTileClick?: (tile: ITile) => void;
  onTagClick: (_event: unknown, value: string) => void;
}

export const BaseCatalog: FunctionComponent<BaseCatalogProps> = (props) => {
  const catalogBodyRef = useRef<HTMLDivElement>(null);
  const itemCount = props.tiles?.length;
  const [page, setPage] = useState(1);
  const [perPage, setPerPage] = useState(50);

  const onTileClick = useCallback(
    (tile: ITile) => {
      props.onTileClick?.(tile);
    },
    [props],
  );

  const startIndex = (page - 1) * perPage < 0 ? 0 : (page - 1) * perPage;
  const endIndex = page * perPage;
  useEffect(() => {
    // Handling the scenario where the item count is less the page selected.
    if (startIndex + 1 > itemCount) {
      setPage(Math.ceil(itemCount / perPage));
    } else if (page === 0 && itemCount > 0) setPage(1);

    catalogBodyRef.current!.scrollTop = 0;
  }, [props.tiles]);

  const onSelectDataListItem = useCallback(
    (_event: React.MouseEvent | React.KeyboardEvent, id: string) => {
      const tile = props.tiles.find((tile) => tile.name + '-' + tile.type === id);
      onTileClick(tile!);
    },
    [onTileClick, props.tiles],
  );

  const onSetPage = (_event: React.MouseEvent | React.KeyboardEvent | MouseEvent, newPage: number) => {
    setPage(newPage);
  };

  const onPerPageSelect = (
    _event: React.MouseEvent | React.KeyboardEvent | MouseEvent,
    newPerPage: number,
    newPage: number,
  ) => {
    setPerPage(newPerPage);
    setPage(newPage);
  };

  const paginatedCards = props.tiles?.slice(startIndex, endIndex);

  return (
    <>
      <Pagination
        itemCount={itemCount}
        perPage={perPage}
        page={page}
        onSetPage={onSetPage}
        widgetId="catalog-pagination"
        onPerPageSelect={onPerPageSelect}
        ouiaId="CatalogPagination"
      />
      <div id="catalog-list" className="catalog-list" ref={catalogBodyRef}>
        {props.catalogLayout == CatalogLayout.List && (
          <DataList aria-label="Catalog list" onSelectDataListItem={onSelectDataListItem} isCompact>
            {paginatedCards.map((tile) => (
              <CatalogDataListItem key={`${tile.name}-${tile.type}`} tile={tile} onTagClick={props.onTagClick} />
            ))}
          </DataList>
        )}
        {props.catalogLayout == CatalogLayout.Gallery && (
          <Gallery hasGutter>
            {paginatedCards.map((tile) => (
              <Tile key={`${tile.name}-${tile.type}`} tile={tile} onClick={onTileClick} onTagClick={props.onTagClick} />
            ))}
          </Gallery>
        )}
      </div>
    </>
  );
};
