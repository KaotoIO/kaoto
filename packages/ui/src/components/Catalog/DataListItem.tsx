import {
  DataListCell,
  DataListItem,
  DataListItemCells,
  DataListItemRow,
  Grid,
  GridItem,
  LabelGroup,
} from '@patternfly/react-core';
import { FunctionComponent } from 'react';
import { IconResolver } from '../IconResolver';
import { ITile } from './Catalog.models';
import './DataListItem.scss';
import { CatalogTag, CatalogTagsPanel } from './Tags';

interface ICatalogDataListItemProps {
  tile: ITile;
  onTagClick: (_event: unknown, value: string) => void;
}

const titleElementOrder = {
  default: '1',
};

const tagsElementOrder = {
  default: '3',
  md: '2',
};

const descriptionElementOrder = {
  default: '2',
  md: '3',
};

export const CatalogDataListItem: FunctionComponent<ICatalogDataListItemProps> = (props) => {
  return (
    <DataListItem
      aria-labelledby={props.tile.name}
      key={props.tile.name}
      id={props.tile.name}
      className="catalog-data-list-item"
    >
      <DataListItemRow>
        <DataListItemCells
          dataListCells={[
            <DataListCell key="List item">
              <Grid>
                <GridItem sm={12} md={6} order={titleElementOrder}>
                  <div className="catalog-data-list-item__title-div-left">
                    <IconResolver className="catalog-data-list-item__title-div-left__icon" tile={props.tile} />
                    <span id="clickable-action-item1" className="catalog-data-list-item__title-div-left__title">
                      {props.tile.title}
                    </span>
                    <LabelGroup isCompact aria-label="data-list-item-headers-tags">
                      {props.tile.headerTags?.map((tag, index) => (
                        <CatalogTag
                          key={`${props.tile.name}-${tag}-${index}`}
                          tag={tag}
                          className="catalog-data-list-item__tags"
                        />
                      ))}
                      {props.tile.version && (
                        <CatalogTag
                          key={`${props.tile.version}`}
                          tag={props.tile.version}
                          className="catalog-data-list-item__tags"
                          variant="outline"
                        />
                      )}
                    </LabelGroup>
                  </div>
                </GridItem>
                <GridItem sm={12} md={6} order={tagsElementOrder}>
                  <div className="catalog-data-list-item__title-div-right">
                    <CatalogTagsPanel tags={props.tile.tags} onTagClick={props.onTagClick} />
                  </div>
                </GridItem>
                <GridItem span={12} order={descriptionElementOrder}>
                  <span className="catalog-data-list-item__description">{props.tile.description}</span>
                </GridItem>
              </Grid>
            </DataListCell>,
          ]}
        />
      </DataListItemRow>
    </DataListItem>
  );
};
