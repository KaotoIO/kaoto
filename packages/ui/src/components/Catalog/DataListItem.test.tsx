import { fireEvent, render } from '@testing-library/react';
import { ITile } from './Catalog.models';
import { CatalogDataListItem } from './DataListItem';

describe('DataListItem', () => {
  const tile: ITile = {
    type: 'tile-type',
    name: 'tile-name',
    title: 'tile-title',
    description: 'tile-description',
    tags: ['tag1', 'tag2'],
    headerTags: ['header-tag1', 'header-tag2'],
    version: '1.0',
  };

  it('renders correctly', () => {
    const { container } = render(<CatalogDataListItem key={tile.name} tile={tile} onTagClick={jest.fn()} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('calls onTagClick prop when clicked', () => {
    const onTagClick = jest.fn();

    const { getByTestId } = render(<CatalogDataListItem key={tile.name} tile={tile} onTagClick={onTagClick} />);

    fireEvent.click(getByTestId('tag-tag1'));

    expect(onTagClick).toHaveBeenCalledTimes(1);
  });
});
