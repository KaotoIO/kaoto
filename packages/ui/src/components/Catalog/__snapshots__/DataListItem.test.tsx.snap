// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataListItem renders correctly 1`] = `
<li
  aria-labelledby="tile-name"
  class="pf-v5-c-data-list__item catalog-data-list-item"
  id="tile-name"
>
  <div
    class="pf-v5-c-data-list__item-row"
  >
    <div
      class="pf-v5-c-data-list__item-content"
    >
      <div
        class="pf-v5-c-data-list__cell"
      >
        <div
          class="pf-v5-l-grid"
        >
          <div
            class="pf-v5-l-grid__item pf-m-12-col-on-sm pf-m-6-col-on-md"
            style="--pf-v5-l-grid--item--Order: 1;"
          >
            <div
              class="catalog-data-list-item__title-div-left"
            >
              <img
                alt="camel icon"
                class="catalog-data-list-item__title-div-left__icon"
                src=""
              />
              <span
                class="catalog-data-list-item__title-div-left__title"
                id="clickable-action-item1"
              >
                tile-title
              </span>
              <div
                class="pf-v5-c-label-group"
              >
                <div
                  class="pf-v5-c-label-group__main"
                >
                  <ul
                    aria-label="data-list-item-headers-tags"
                    class="pf-v5-c-label-group__list"
                    role="list"
                  >
                    <li
                      class="pf-v5-c-label-group__list-item"
                    >
                      <span
                        class="pf-v5-c-label pf-m-compact catalog-data-list-item__tags"
                      >
                        <span
                          class="pf-v5-c-label__content"
                        >
                          <span
                            class="pf-v5-c-label__text"
                          >
                            header-tag1
                          </span>
                        </span>
                      </span>
                    </li>
                    <li
                      class="pf-v5-c-label-group__list-item"
                    >
                      <span
                        class="pf-v5-c-label pf-m-compact catalog-data-list-item__tags"
                      >
                        <span
                          class="pf-v5-c-label__content"
                        >
                          <span
                            class="pf-v5-c-label__text"
                          >
                            header-tag2
                          </span>
                        </span>
                      </span>
                    </li>
                    <li
                      class="pf-v5-c-label-group__list-item"
                    >
                      <span
                        class="pf-v5-c-label pf-m-outline pf-m-compact catalog-data-list-item__tags"
                      >
                        <span
                          class="pf-v5-c-label__content"
                        >
                          <span
                            class="pf-v5-c-label__text"
                          >
                            1.0
                          </span>
                        </span>
                      </span>
                    </li>
                  </ul>
                </div>
              </div>
            </div>
          </div>
          <div
            class="pf-v5-l-grid__item pf-m-12-col-on-sm pf-m-6-col-on-md"
            style="--pf-v5-l-grid--item--Order: 3; --pf-v5-l-grid--item--Order-on-md: 2;"
          >
            <div
              class="catalog-data-list-item__title-div-right"
            >
              <div
                class="pf-v5-c-label-group"
              >
                <div
                  class="pf-v5-c-label-group__main"
                >
                  <ul
                    aria-label="data-list-item-tags"
                    class="pf-v5-c-label-group__list"
                    role="list"
                  >
                    <li
                      class="pf-v5-c-label-group__list-item"
                    >
                      <span
                        class="pf-v5-c-label pf-m-compact"
                        data-testid="tag-tag1"
                      >
                        <a
                          class="pf-v5-c-label__content"
                        >
                          <span
                            class="pf-v5-c-label__text"
                          >
                            tag1
                          </span>
                        </a>
                      </span>
                    </li>
                    <li
                      class="pf-v5-c-label-group__list-item"
                    >
                      <span
                        class="pf-v5-c-label pf-m-compact"
                        data-testid="tag-tag2"
                      >
                        <a
                          class="pf-v5-c-label__content"
                        >
                          <span
                            class="pf-v5-c-label__text"
                          >
                            tag2
                          </span>
                        </a>
                      </span>
                    </li>
                  </ul>
                </div>
              </div>
            </div>
          </div>
          <div
            class="pf-v5-l-grid__item pf-m-12-col"
            style="--pf-v5-l-grid--item--Order: 2; --pf-v5-l-grid--item--Order-on-md: 3;"
          >
            <span
              class="catalog-data-list-item__description"
            >
              tile-description
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</li>
`;
